tiatoolbox/tools/patchextraction.py

```diff
@@ -53,29 +52,86 @@ class PatchExtractor(ABC):
         input_img(ndarray, WSIReader): input image for patch extraction.
           input_image type is ndarray for an image tile whereas :obj:`WSIReader`
           for an WSI.
-        patch_size(Tuple of int): patch size tuple (width, height).
-        resolution(Tuple of int): resolution at which to read the image.
+        patch_size(tuple(int)): patch size tuple (width, height).
+        resolution(tuple(int)): resolution at which to read the image.
         units (str): the units of resolution.
         n(int): current state of the iterator.
+        locations_df(pd.DataFrame): A table containing location and/or type of patch.
 
     """
 
     def __init__(self, input_img, patch_size, resolution=0, units="level"):
-        self.patch_size = patch_size
+        if isinstance(patch_size, (tuple, list)):
+            self.patch_size = (int(patch_size[0]), int(patch_size[1]))
+        else:
+            self.patch_size = (int(patch_size), int(patch_size))
         self.resolution = resolution
         self.units = units
         self.n = 0
         self.wsi = get_wsireader(input_img=input_img)
+        self.locations_df = None
+        self.stride = None
 
     def __iter__(self):
         self.n = 0
         return self
 
     def __next__(self):
-        raise NotImplementedError
+        n = self.n
+
+        if n >= self.locations_df.shape[0]:
+            raise StopIteration
+        self.n = n + 1
+        return self[n]
 
     def __getitem__(self, item):
-        raise NotImplementedError
+        if type(item) is not int:
+            raise TypeError("Index should be an integer.")
+
+        if item >= self.locations_df.shape[0]:
+            raise IndexError
+
+        x = self.locations_df["x"][item]
+        y = self.locations_df["y"][item]
+
+        data = self.wsi.read_rect(
+            location=(int(x), int(y)),
+            size=self.patch_size,
+            resolution=self.resolution,
+            units=self.units,
+        )
+
+        return data
+
+    def _generate_location_df(self):
+        """Generate location list based on slide dimension.
+        The slide dimension is calculated using units and resolution.
+
+        """
+        level, _ = self.wsi.find_optimal_level_and_downsample(
+            resolution=self.resolution, units=self.units
+        )
+
+        slide_dimension = self.wsi.info.level_dimensions[level]
+
+        img_w = slide_dimension[0]
+        img_h = slide_dimension[1]
+        img_patch_w = self.patch_size[0]
+        img_patch_h = self.patch_size[1]
+        stride_w = self.stride[0]
+        stride_h = self.stride[1]
+
+        data = []
+
+        for h in range(int(math.ceil((img_h - img_patch_h) / stride_h + 1))):
+            for w in range(int(math.ceil((img_w - img_patch_w) / stride_w + 1))):
+                start_h = h * stride_h
+                start_w = w * stride_w
+                data.append([start_w, start_h, None])
```

Comment:
You may want to use np.meshgrid to optimize this

Reply:
I am leaving this for a separate PR as this needs to be carefully checked.